stablehlo/dialect/TypeInference.cpp

```diff
@@ -1223,7 +1223,35 @@ static LogicalResult inferGatherReturnTypeComponents(
                             offsetDims, collapsedSliceDims, startIndexMap,
                             indexVectorDim, shape);
 
-  inferredReturnShapes.emplace_back(shape, elementType);
+  // The dimension sizes of result, corresponding to offset dimensions, depend
```

Comment:
Hi Smit,

That would be nice and I considered the same.  PTAL at the PR description where I noted my findings. Let me know if I am missing anything.

Reply:
ah, I see. sorry, didn't read the description.

I personally would avoid the duplication as this shouldn't negatively affect the other user in practice and will be easier to follow.

Reply:
How about we apply and patch this PR in MHLO, if everything is OK then we can go on with shared code among `reifyReturnTypeShapes` with confidence?

Reply:
That sounds good. 

@smit-hinsu @zhouxin913 
Sharing the code might still have problems.

The main reason I did not share [the](https://github.com/openxla/stablehlo/blob/afcbb953c12a2a3a8564cab73278f4b47021ec01/stablehlo/dialect/TypeInference.cpp#L1119) between  `GatherOp::inferReturnTypeComponents` and `GatherOp::reifyReturnTypeShapes` is because I do not think bounds are currently handled in reify functions. For example, which `op` to create to reify bounds from input arguments.

Reply:
We don't need to reify bounds in reifyReturnTypeShapes.

Can't we just use inferGatherShape to infer shape and bounds. Use both in inferReturnTypeComponents and drop it within reifyGatherDimSizes. (I don't have the full context around reify.)

Reply:
Hi Smit,

Here is what I found to be done if we want to share  `inferGatherShape` among  ` reifyGatherDimSizes` and `inferReturnTypeComponents`

Both ` reifyGatherDimSizes` and `inferReturnTypeComponents` need to provide callbacks (similar to `getStartIndicesDim` and `getSliceDim`) which, for `inferReturnTypeComponents`, will help compute the concrete bounds and, for `reifyGatherDimSizes`, help reify the bounds.

The callbacks for computing the concrete bounds are straightforward. Something like
```
  ArrayRef<int64_t> startIndicesBounds = encodingToBounds(
      startIndices.getType().cast<RankedTensorType>().getEncoding());
  
auto getOffsetIndicesBound = [&](int64_t index) {
    return ShapedType::kDynamic;
  };

  auto  getBatchIndicesBound = [&](int64_t index) {
    if(startIndicesBounds.empty())
      return ShapedType::kDynamic;
    return startIndicesBounds[index];
  };

```

For for reifying the bounds, I need to pass some dummy callbacks which will end up creating some dead `mlir::Value` typed objects, probably of type `ConstantIndexOp`, using builder API.

As we are not sure about the full context around reify, my proposal is to keep them separate for now. With more context (which IMO will arise once we have a need to support bounds in reify methods) we can make an  effort to clean this tech debt.

Reply:
That sounds completely reasonable. Thanks for digging in!

The code duplication is not large anyway.

Reply:
Thanks!